### PR TITLE
Adding dnsConfig for test pods

### PR DESCRIPTION
**Pull request by deepsm007**
Adding the DNSCONFIG for pod spec [DPTP-1260](https://issues.redhat.com/browse/DPTP-1260)

Signed-off-by: Deep Mistry

### CODE REVIEW

**Comment by petr-muller** [comment]
IMO this should be a card on DPTP board first to discuss this before the implementation. We generally do not want to expose Podspec members directly in steps so this goes against the design.

/hold

**Comment by petr-muller** [comment]
/hold cancel

The discussion I asked for actualy happened DPTP-1260

**Comment by alvaroaleman** [comment]
@deepsm007 it seems something went wrong when you rebased, the only commit in this PR is a mergecommit and you have attached your changes to that, please fix it to be a proper, non-merge commit.
/hold

File: pkg/api/types.go [hunk 1]
```diff
@@ -628,6 +628,8 @@ type LiteralTestStep struct {
 	// Dependencies lists images which must be available before the test runs
 	// and the environment variables which are used to expose their pull specs.
 	Dependencies []StepDependency `json:"dependencies,omitempty"`
+	// DnsConfig for step's Pod.
+	DNSConfig StepDNSConfig `json:"dnsConfig,omitempty"`
```
Comment (alvaroaleman) [inline pkg/api/types.go:632]: Lets make all the DNSConfig fields pointers to avoid serializing it into all config

File: pkg/steps/multi_stage.go [hunk 2]
```diff
@@ -430,6 +430,9 @@ func (s *multiStageTestStep) generatePods(steps []api.LiteralTestStep, env []cor
 		pod.Labels[MultiStageTestLabel] = s.name
 		pod.Spec.ServiceAccountName = s.name
 		pod.Spec.TerminationGracePeriodSeconds = terminationGracePeriodSeconds
+		for idx := range step.DNSConfig.Searches {
+			pod.Spec.DNSConfig.Searches = append(pod.Spec.DNSConfig.Searches, step.DNSConfig.Searches[idx])
+		}
```
Comment (petr-muller) [inline pkg/steps/multi_stage.go:435]: No need to iterate:

```
pod.Spec.DNSConfig.Searches = append(pod.Spec.DNSConfig.Searches, step.DNSConfig.Searches...)
```

File: pkg/validation/test_test.go [hunk 3]
```diff
@@ -990,6 +990,44 @@ func TestValidateDependencies(t *testing.T) {
...
+			},
+			output: []error{
+				errors.New("root.searches[0] must be set"),
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			if actual, expected := validateDNSConfig("root", testCase.input), testCase.output; !reflect.DeepEqual(actual, expected) {
```
Comment (petr-muller) [inline pkg/validation/test_test.go:1022]: You can avoid `reflect.DeepEqual` by only using `cmp.Diff`, resulting in simpler code:

```
if diff := cmp.Diff(one, two, comparer); diff != "" {
t.Errorf("...%s", diff)
}
```

File: pkg/api/types.go [hunk 4]
```diff
@@ -760,6 +771,8 @@ type MultiStageTestConfigurationLiteral struct {
 	Environment TestEnvironment `json:"env,omitempty"`
 	// Dependencies holds override values for dependency parameters.
 	Dependencies TestDependencies `json:"dependencies,omitempty"`
+	// DnsConfig for step's Pod.
+	DNSConfig *StepDNSConfig `json:"dnsConfig,inline,omitempty"`
```
Comment (alvaroaleman) [inline pkg/api/types.go:775]: Why `inline`?

File: pkg/steps/multi_stage.go [hunk 5]
```diff
@@ -430,6 +430,12 @@ func (s *multiStageTestStep) generatePods(steps []api.LiteralTestStep, env []cor
 		pod.Labels[MultiStageTestLabel] = s.name
 		pod.Spec.ServiceAccountName = s.name
 		pod.Spec.TerminationGracePeriodSeconds = terminationGracePeriodSeconds
+		if step.DNSConfig != nil {
```
Comment (alvaroaleman) [inline pkg/steps/multi_stage.go:433]: Can you please add a test where this is exercised?

Comment (deepsm007) [reply pkg/steps/multi_stage.go:433]: Yes, i am working on the test.

**Review by alvaroaleman — APPROVED** [review]
/label tide/merge-method-squash

**Pull request merged**